cool-core/src/main/java/com/nus/cool/core/cohort/refactor/CohortProcessor.java

```diff
@@ -109,14 +107,17 @@ private void processCublet(CubletRS cublet) {
         // we only get used schema;
         for (String schema : this.projectedSchemaSet) {
             MetaFieldRS metaField = metaChunk.getMetaField(schema);
-            if (FieldType.IsHashType(metaField.getFieldType())) {
+            if(FieldType.IsUserKeyField(metaField.getFieldType())){
+                gidMapBySchema.put(schema,((UserMetaFieldRS)metaField).getGidMap());
+            }
+            else if (FieldType.IsHashType(metaField.getFieldType())) {
                 gidMapBySchema.put(schema, ((HashMetaFieldRS) metaField).getGidMap());
             }
         }
 
         // Now start to pass the DataChunk
         for (ChunkRS chunk : cublet.getDataChunks()) {
-            this.processDataChunk(chunk, gidMapBySchema);
+            this.processDataChunk(chunk, gidMapBySchema, ((UserMetaFieldRS)(metaChunk.getMetaField(0,FieldType.UserKey))));
```

Comment:
It should be a method of MetaChunkRS getUserMetaField, so we do not have this explicit type cast here.